### PR TITLE
dflash: AngelSpec DFly drafter support

### PR DESCRIPTION
**Pull request by bri-prism**
## What

Adds AngelSpec DFly drafter support to the `dflash` arch, so a DFly checkpoint can be
converted and used as a speculative drafter. Verified against the public reference pairing:
Qwen3-8B as target and `AngelSlim/Qwen3-8B-DFly-Block8` as drafter.

Also fixes one bug in `common/speculative.cpp` that is independent of DFly and could be
taken on its own (second commit).

## Why

DFly extends the DFlash draft path with two pieces we did not have:

1. **Per-draft-layer context fusion.** DFlash projects the captured target features once and
   injects that single context into every draft layer. DFly keeps that shared projection as a
   base context, then adds a per-draft-layer softmax-weighted mix of the raw per-target-layer
   features, so each draft layer gets its own context.
2. **A TreeFlash predecessor correction** applied before the target head, chained across the
   block: position i's logits come from the draft hidden state at row i corrected by the
   embedding of the token drafted at row i-1.

The expensive prerequisite was already here. `dflash` already reads a `target_layers` array
and sizes its encoder input as `|target_layers| * n_embd`, which is exactly the concatenated
feature layout DFly consumes, so multi-layer capture needed no work.

## How

DFly is detected from the presence of the `layer_fusion` tensor rather than from a metadata
key, so a DFly export cannot quietly load as plain DFlash. A checkpoint that declares both a
DFly fusion and a DSpark Markov head is rejected with that reason instead of failing later on
a missing tensor. Reduced draft vocabularies are rejected, since the reference chain runs the
full target head.

Nothing was needed in the draft reader. DFly reports no Markov head, so it already takes the
DFlash path (block rows 1 through n-1) and the plain per-token probability cut.

Two width contracts have to widen together for the fused context to survive its round trip
through the host, and this is the part worth reviewing:

* `n_embd_out` sizes the encoder's nextn output. `common/speculative.cpp` was sizing every
  buffer on that path with `llama_model_n_embd(model_dft)` instead. Those two are equal for
  any drafter that emits one context row per token, so that commit is a no-op for DFlash and
  DSpark, but they are not equal for DFly and the batch was under-allocated.
* `n_embd_inp` sizes the embd batch that feeds the context back in. A dflash draft context is
  not MTP-typed, so `llama_context::decode` takes the `n_embd_inp()` branch. Setting only
  `n_embd_out` leaves the decoder reading one layer's context and n_layer-1 layers of junk.

## Measurements

M5 Pro, Metal, bf16 target and bf16 drafter, block 8, greedy, 64 predicted tokens, through
`llama-server`, with the arms interleaved A-B-A-B:

| arm | tok/s | acceptance |
|---|---|---|
| autoregressive baseline | 18.35, 18.42, 18.50, 18.52 | n/a |
| DFly | 38.53, 38.65, 38.18, 38.00 | 52.7 percent (49/93) |
| DFly, correction disabled | 30.89 | 21.0 percent (37/176) |

That is 2.08x against a same-path baseline whose four readings spread 0.9 percent. Mean
accepted length is 4.50 of a block of 8, which puts an 8-row verify at about 2.16 plain decode
steps. That is the right order for a bandwidth-bound bf16 8B target, so the mechanism accounts
for the size of the speedup rather than just accompanying it.

The correction is load-bearing: disabling it drops acceptance from 52.7 to 21.0 percent.
`LLAMA_DFLY_NO_CHAIN=1` exists to make that A/B measurable and is not a normal setting.

## Testing

* Generated text is byte identical to the autoregressive baseline at greedy in every
  speculative arm. That is the guarantee speculative decoding owes, and it is what would catch
  a mis-indexed block or a corrupted verify.
* `tests/test-dfly-fusion.cpp` pins the fusion math and the resulting memory layout against an
  independent scalar reference. The axis handling needs two reshape and permute round trips,
  and a transposed axis would otherwise still load and still draft plausibly. The test also
  asserts the per-layer contexts actually differ, so a collapsed broadcast fails.
* `tests/gen-tiny-dfly.py` builds a tiny random DFly GGUF for loader smoke tests, including
  the rejection paths above.

## Notes for review

The conversion path covers `Qwen3DFlyModel` and the `Qwen3DSparkDFlareV2Model` alias, and
refuses a config whose `target_layer_ids` exceed its stated target depth. The published
metadata of the reference checkpoint does exactly that on its main branch, and it loads clean
and then mis-projects, so pin a known good revision when converting.

### CODE REVIEW

File: conversion/qwen.py [hunk 1]
```diff
@@ -729,6 +729,83 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
...
+        super().__init__(dir_model, *args, hparams=hparams, **kwargs)
+
+        hp = self.hparams
+
+        # The published main-branch config of the reference checkpoint states a target depth and
+        # vocab that do not match the weights (80/120832 against the real 36/151936), which loads
+        # clean and then mis-projects. Refuse instead of converting a checkpoint that lies.
+        target_layers = hp.get("target_num_hidden_layers")
+        layer_ids     = hp.get("target_layer_ids") or []
+        if target_layers and layer_ids and max(layer_ids) >= int(target_layers):
```
[the remark this answers is not in the document]
Comment (bri-prism) [reply conversion/qwen.py:778]: Fixed in `920c50d1d`. You were right that the old bound could not catch the case it was written for: the reference config's declared 80 is self-consistent with capture ids up to 33, so a declared depth cannot police itself.

The declared `target_num_hidden_layers`, `target_vocab_size` and `target_hidden_size` are now each compared against the target model's own `config.json`, read from `--target-model-dir`, and the ids are bounded by the real depth rather than the declared one. Nested configs are followed with the same `text_config` merge the rest of this file uses.

I checked this both ways rather than asserting it: the 80/120832 config against a real 36/151936 target is accepted by the pre-fix code and rejected after, and a truthful config still converts. Shapes are read per key, so one unusable value does not disable the other comparisons, and a target config that is missing, unparseable or not a JSON object leaves the comparison inert instead of raising here, since `set_vocab` reads the same file and fails on it anyway. Without `--target-model-dir` the cross-check is skipped and the old bound still applies.

File: conversion/qwen.py [hunk 2]
```diff
@@ -729,6 +729,83 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
...
+                "found; the export is incomplete and would draft without the correction."
+            )
+
+    _seen_correction = False
+
+    def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iterable[tuple[str, Tensor]]:
+        if "hidden_correction." in name:
+            self._seen_correction = True
+
+        yield from super().modify_tensors(data_torch, name, bid)
```
[the remark this answers is not in the document]
Comment (bri-prism) [reply conversion/qwen.py:852]: Fixed in `920c50d1d`. `hidden_correction.*` tensors are now omitted when `enable_hidden_correction` is false, with one log line saying so rather than dropping them silently.

Worth noting for the record that there is no metadata key to fix alongside this. Tensor presence is the only signal the runtime has, so omitting the tensors is the whole fix.

Behaviour was checked against the real class rather than argued: with correction disabled the path yields no tensors and the seen flag stays clear, and with it enabled the tensor and every unrelated tensor behave identically with and without the change. Only the disabled case differs.

File: src/models/dflash.cpp [hunk 3]
```diff
@@ -389,6 +480,138 @@ static void build_dspark_markov_head(llm_graph_context & g, const llama_model &
...
+                NULL, LLM_FFN_SILU, LLM_FFN_PAR, -1);
+
+        ggml_tensor * col = g.build_lora_mm(output, ggml_add(ctx0, h_i, delta), output_s);
+
+        cat = ggml_concat(ctx0, cat, col, 1);
+
+        // greedy chain: the next position conditions on this position's drafted token
+        if (i + 1 < block_drafts) {
+            prev = ggml_argmax(ctx0, col);
+        }
```
[the remark this answers is not in the document]
Comment (bri-prism) [reply src/models/dflash.cpp:613]: I went looking for this one expecting to have to fix it, and it does not hold. The driver never uses the token the sampler chain selects.

In `common_speculative_impl_draft_dflash::update`, both the DSpark and the DFlash branches do the same thing: they call `common_sampler_sample(...)` and discard its return value, then take the id from the candidate list, `const llama_token id = cur_p->data[0].id;`. `common_sampler_get_candidates(smpl, true)` sorts the array descending by probability before returning it, so `data[0]` is the highest-probability token, which is the same argmax the graph takes with `ggml_argmax`. The stochastic pick from the `dist` sampler is recorded in `cur_p->selected`, and nothing reads `data[selected]`.

So the token appended to the draft is the argmax of exactly the column the next correction conditions on, and the chain is consistent. `common_sampler_accept` is then called with that same argmax, so the sampler's own history does not drift either. This holds on the backend-sampling path too: the device chain only pre-filters to the top 10, and the final choice is still `data[0]` in the same loop, so there is no CPU versus backend divergence to reconcile.

What is fair in the observation is that the code reads as if it samples. The chain is `top_k(10)` followed by the default `dist`, so a chain that looks stochastic is built, one token is drawn from it, and that draw is thrown away, while `top_k` is 10 when only the first entry is ever consumed. That is wasted work and it is why this looks wrong on a read of the setup rather than the consumption.

I would rather make the intent explicit than leave this to be re-raised, so I will follow up separately with the greedy selection stated at the point of choice and the candidate breadth reduced to what is actually used. That is a readability and small-cost change with no effect on which token is drafted, which keeps it clear of this PR's measured acceptance numbers. Resolving this thread since the reported defect is not present.

**Comment by bri-prism** [comment]
Superseded by #172, re-branched off `prism`. This PR targeted `prism-v7`, which is frozen, and changing the base in place would pull in the two `prism-v7`-only commits sitting under the branch (`555c5cda2` Hadamard activation sharing, #137, and `3c12dbd5f` PQ2_0 ftype naming, #134) rather than just retargeting it. Both are already on `prism` under different SHAs.

All seven commits cherry-picked onto `prism` with no conflicts, same 13 files and +816/-15. Eleven of the thirteen are byte-identical to `920c50d1d`; `gguf-py/gguf/constants.py` and `tests/CMakeLists.txt` differ only where `prism` carries PTQ1_0 (#148), which `prism-v7` never received.

Carried forward to #172: the three resolved review threads and the Windows test failure from the 2026-09-04 run, which gets a fresh evaluation on the new base.
